### PR TITLE
Add atomic_flush option for checkpoint

### PR DESCRIPTION
**Pull request by xingbowang**
## Summary

Add `force_atomic_flush` to `FlushOptions` so any flush caller (Checkpoint, `DB::Flush()`, etc.) can force atomic flush of all column families even when `DBOptions::atomic_flush` is disabled.

### Public API
- Added `bool force_atomic_flush = false` to `FlushOptions`
- Added `bool atomic_flush = false` to `LiveFilesStorageInfoOptions`

### Flush dispatch
- `FlushAllColumnFamilies()` — uses `immutable_db_options_.atomic_flush || flush_options.force_atomic_flush`
- `DB::Flush()` (single-CF and multi-CF) — respects `force_atomic_flush`
- `FlushForGetLiveFiles()` — constructs `FlushOptions` with `force_atomic_flush` from `LiveFilesStorageInfoOptions::atomic_flush`

### Per-request atomic flush in pipeline
- `FlushRequest.atomic_flush` and `BGFlushArg.atomic_flush_` carry the per-request flag
- `GenerateFlushRequest`, `EnqueuePendingFlush`, `PopFirstFromFlushQueue`, `BackgroundFlush`, `FlushMemTablesToOutputFiles` all use the per-request flag

### Stress test
- Added `--checkpoint_atomic_flush` flag to db_stress, randomly enabled in db_crashtest.py

### Unit tests (6 new)
- End-to-end checkpoint with atomic flush override
- Negative test (default = no atomic flush)
- Single-CF edge case
- Interaction with `DBOptions::atomic_flush=true`
- Mixed non-atomic then atomic flushes in queue
- Mixed atomic then non-atomic flushes in queue

### CODE REVIEW

File: db/db_impl/compacted_db_impl.h [hunk 1]
```diff
@@ -129,7 +129,7 @@ class CompactedDBImpl : public DBImpl {
   // Share with DBImplReadOnly?
 
  protected:
-  Status FlushForGetLiveFiles() override {
+  Status FlushForGetLiveFiles(bool /*force_atomic_flush*/ = false) override {
```
Comment (xingbowang) [inline db/db_impl/compacted_db_impl.h]: let's remove the default value, update the call site to always pass in the value.

File: db/db_impl/db_impl.h [hunk 2]
```diff
@@ -1529,9 +1529,10 @@ class DBImpl : public DB {
       ColumnFamilyData* cfd, const ExternalSstFileIngestionJob& ingestion_job);
 
   Status FlushAllColumnFamilies(const FlushOptions& flush_options,
-                                FlushReason flush_reason);
+                                FlushReason flush_reason,
+                                bool force_atomic_flush = false);
```
Comment (xingbowang) [inline db/db_impl/db_impl.h]: Same here. Remove the default argument and always pass the value in the call site.

File: db/db_impl/db_impl.h [hunk 3]
```diff
@@ -1915,11 +1916,12 @@ class DBImpl : public DB {
           flush_reason_(FlushReason::kOthers) {}
     BGFlushArg(ColumnFamilyData* cfd, uint64_t max_memtable_id,
                SuperVersionContext* superversion_context,
-               FlushReason flush_reason)
+               FlushReason flush_reason, bool atomic_flush = false)
```
Comment (xingbowang) [inline db/db_impl/db_impl.h]: Same as above.

File: db/db_impl/db_impl.h [hunk 4]
```diff
@@ -1529,9 +1529,10 @@ class DBImpl : public DB {
       ColumnFamilyData* cfd, const ExternalSstFileIngestionJob& ingestion_job);
 
   Status FlushAllColumnFamilies(const FlushOptions& flush_options,
-                                FlushReason flush_reason);
+                                FlushReason flush_reason,
+                                bool force_atomic_flush);
```
Comment (joshkang97) [inline db/db_impl/db_impl.h]: Would it be cleaner to put `force_atomic_flush` in FlushOptions. Then callers can also utilize this in DB->Flush()

File: utilities/checkpoint/checkpoint_test.cc [hunk 5]
```diff
@@ -1086,6 +1087,168 @@ TEST_P(CheckpointDestroyTest, DisableEnableSlowDeletion) {
...
+  ASSERT_TRUE(atomic_flush_called);
+
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->ClearAllCallBacks();
+
+  ASSERT_EQ("val0", Get(0, "key0"));
+  ASSERT_EQ("val1", Get(1, "key1"));
+  ASSERT_EQ("val2", Get(2, "key2"));
+}
+
```
Comment (joshkang97) [inline utilities/checkpoint/checkpoint_test.cc:1251]: should we add a test where there is both a non-atomic and atomic flush in the flush queue

File: db/db_impl/compacted_db_impl.h (hunk 1 above)
Comment (xingbowang) [reply db/db_impl/compacted_db_impl.h]: Done. In the latest push, this parameter has been removed entirely — `force_atomic_flush` now lives in `FlushOptions` instead (per your other comment).

File: db/db_impl/db_impl.h (hunk 2 above)
Comment (xingbowang) [reply db/db_impl/db_impl.h]: Done — `force_atomic_flush` is now a field in `FlushOptions`, so this separate parameter is gone entirely.

File: db/db_impl/db_impl.h (hunk 3 above)
Comment (xingbowang) [reply db/db_impl/db_impl.h]: Done.

File: db/db_impl/db_impl.h (hunk 4 above)
Comment (xingbowang) [reply db/db_impl/db_impl.h]: Great idea — done. `force_atomic_flush` is now a field in `FlushOptions`. Both `DB::Flush()` (single-CF and multi-CF) and the internal `FlushAllColumnFamilies()` respect it. `FlushForGetLiveFiles()` constructs a `FlushOptions` with the flag set from `LiveFilesStorageInfoOptions::atomic_flush`.

File: utilities/checkpoint/checkpoint_test.cc (hunk 5 above)
Comment (xingbowang) [reply utilities/checkpoint/checkpoint_test.cc:1251]: Added two tests covering both orders:
- `MixedAtomicAndNonAtomicFlushInQueue` — enqueues non-atomic first, then atomic
- `MixedAtomicThenNonAtomicFlushInQueue` — enqueues atomic first, then non-atomic

Both pause background work, enqueue both flush types, resume, and verify all data is readable.

**Comment by xingbowang** [comment]
Added `DBFlushWithForceAtomicFlush` test — covers both multi-CF and single-CF `DB::Flush()` with `FlushOptions::force_atomic_flush=true`. Verifies via SyncPoint that `AtomicFlushMemTables` is invoked even when `DBOptions::atomic_flush` is false.

**Comment by xingbowang** [comment]
Added DBFlushWithForceAtomicFlush test — covers both multi-CF and single-CF DB::Flush() with force_atomic_flush=true. Verifies via SyncPoint that AtomicFlushMemTables is invoked even when DBOptions::atomic_flush is false.
